lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -1770,7 +1770,10 @@ def select_datastore(self, vm_obj=None):
             # TODO: really use the datastore for newly created disks
             if 'autoselect_datastore' in self.params['disk'][0] and self.params['disk'][0]['autoselect_datastore']:
                 datastores = self.cache.get_all_objs(self.content, [vim.Datastore])
-                datastores = [x for x in datastores if self.cache.get_parent_datacenter(x).name == self.params['datacenter']]
+                if self.params['esxi_hostname'] is not None:
```

Comment:
Since for Python, a non-empy string is True, you can just do: `if self.params['esxi_hostname']:`.